rulesets/owasp_ruleset_functions.go

```diff
@@ -0,0 +1,893 @@
+		Description:  "Check if the operation is protected at operation level.\nOtherwise, check the global `#/security` property",
+		Given:        `$`,
+		Resolved:     false,
+		RuleCategory: model.RuleCategories[model.CategoryInfo],
+		Recommended:  true,
+		Type:         Validation,
+		Severity:     model.SeverityInfo,
+		Then: model.RuleAction{
+			Function: "owaspCheckSecurity",
+			FunctionOptions: map[string]interface{}{
+				"schemesPath": []string{"securitySchemes"},
+				"nullable":    false,
+				"methods":     []string{"post", "put", "patch", "delete"},
+			},
+		},
+	}
+}
+
+// https://github.com/italia/api-oas-checker/blob/master/security/security.yml
+func GetOWASPProtectionGlobalSafeRule() *model.Rule {
+	return &model.Rule{
+		Name:         "This operation is not protected by any security scheme",
+		Id:           OwaspProtectionGlobalSafe,
+		Description:  "Check if the operation is protected at operation level.\nOtherwise, check the global `#/security` property",
+		Given:        `$`,
+		Resolved:     false,
+		RuleCategory: model.RuleCategories[model.CategoryInfo],
+		Recommended:  true,
+		Type:         Validation,
+		Severity:     model.SeverityInfo,
+		Then: model.RuleAction{
+			Function: "owaspCheckSecurity",
+			FunctionOptions: map[string]interface{}{
+				"schemesPath": []string{"securitySchemes"},
+				"nullable":    true,
+				"methods":     []string{"get", "head"},
+			},
+		},
+	}
+}
+
+func GetOWASPDefineErrorValidationRule() *model.Rule {
+
+	return &model.Rule{
+		Name:         "Missing error response of either 400, 422 or 4XX",
+		Id:           OwaspDefineErrorValidation,
+		Description:  "Carefully define schemas for all the API responses, including either 400, 422 or 4XX responses which describe errors caused by invalid requests",
+		Given:        `$.paths..responses`,
+		Resolved:     false,
+		RuleCategory: model.RuleCategories[model.CategoryInfo],
+		Recommended:  true,
+		Type:         Validation,
+		Severity:     model.SeverityWarn,
+		Then: model.RuleAction{
+			Function: "owaspDefineErrorDefinition",
+		},
+	}
+}
+
+func GetOWASPDefineErrorResponses401Rule() *model.Rule {
+
+	// create a schema to match against.
+	opts := make(map[string]interface{})
+	yml := `type: object
+required:
+  - content`
+
+	jsonSchema, _ := parser.ConvertYAMLIntoJSONSchema(yml, nil)
+	opts["schema"] = jsonSchema
+	opts["forceValidation"] = true // this will be picked up by the schema function to force validation.
+
+	return &model.Rule{
+		Name:         "Operation is missing {{property}}",
+		Id:           OwaspDefineErrorResponses401,
+		Description:  "OWASP API Security recommends defining schemas for all responses, even errors. The 401 describes what happens when a request is unauthorized, so its important to define this not just for documentation, but to empower contract testing to make sure the proper JSON structure is being returned instead of leaking implementation details in backtraces",
```

Comment:
Can you break this into a message and `HowToFIx`

Reply:
done